src/Moq/CaptureMatch.cs

```diff
@@ -27,30 +42,80 @@ public CaptureMatch(Action<T> captureCallback)
 		/// <param name="captureCallback">An action to run on captured value</param>
 		/// <param name="predicate">A predicate used to filter captured parameters</param>
 		public CaptureMatch(Action<T> captureCallback, Expression<Func<T, bool>> predicate)
-			: base(CreatePredicate(captureCallback, predicate), () => It.Is(predicate))
+			: this(captureCallback, predicate, false)
+		{
+		}
+		
+		/// <summary>
+		/// Initializes an instance of the capture match.
+		/// </summary>
+		/// <param name="captureCallback">An action to run on captured value</param>
+		/// <param name="predicate">A predicate used to filter captured parameters</param>
+		/// <param name="captureOnSuccessOnly">
+		/// Indicates whether <paramref name="captureCallback"/> should be invoked on parameter evaluation
+		/// or only after all parameters were successfully evaluated.
+		/// </param>
+		public CaptureMatch(Action<T> captureCallback, Expression<Func<T, bool>> predicate, bool captureOnSuccessOnly)
+			: this(new Matcher(captureCallback, predicate, captureOnSuccessOnly))
 		{
 		}
 
-		private static Predicate<T> CreatePredicate(Action<T> captureCallback)
+		private CaptureMatch(Matcher matcher)
+			: base(matcher.Evaluate, matcher.RenderExpression)
 		{
-			return value =>
-			{
-				captureCallback.Invoke(value);
-				return true;
-			};
+			this.matcher = matcher;
 		}
 
-		private static Predicate<T> CreatePredicate(Action<T> captureCallback, Expression<Func<T, bool>> predicate)
+		internal override void OnSuccess()
 		{
-			var predicateDelegate = predicate.CompileUsingExpressionCompiler();
-			return value =>
+			this.matcher.OnSuccess();
+		}
+
+		private class Matcher
```

Comment:
I added this type because `CaptureMatch` inherits from `Match` and the current `Match` constructor needs the condition and the renderExpression. I see two other options:
- Add a parameterless constructor to `Match`.
- Invoke the base constructor using `base(null, null)` and then set the properties.
I will add the parameterless constructor.